File: mcp/src/tools/stk-push.ts
import type { Mpesa } from "@daraja-sdk/ts";
import type { Tool } from "./index.js";

export const stkPushTool: Tool = {
  name: "stk_push",
  description: "Initiate an STK Push (Lipa Na M-Pesa Online) to send a USSD prompt to a customer's phone for payment. Amount range: 1-250000. Returns CheckoutRequestID for status queries.",
  inputSchema: {
    businessShortCode: { type: "number", description: "Organization's shortcode (5-7 digits)" },
    amount: { type: "number", description: "Transaction amount (1-250000)" },
    partyA: { type: "string", description: "Phone number sending money (format: 254XXXXXXXXX)" },
    partyB: { type: "string", description: "Organization receiving funds (shortcode or till number)" },
    phoneNumber: { type: "string", description: "Customer phone number to receive STK prompt (format: 254XXXXXXXXX)" },
    accountReference: { type: "string", description: "Account reference (max 12 characters)" },
    transactionDesc: { type: "string", description: "Transaction description (max 13 characters)" },
  },
  handler: async (input, client) => {
    return client.stkPush.initiate({
      BusinessShortCode: input.businessShortCode as number,
      Amount: input.amount as number,
      PartyA: input.partyA as string,
      PartyB: input.partyB as string,
      PhoneNumber: input.phoneNumber as string,
      AccountReference: input.accountReference as string,
      TransactionDesc: input.transactionDesc as string,
      CallBackURL: "",
      TransactionType: "CustomerBuyGoodsOnline",
    } as never);
  },
};

File: examples/go/stk_push.go
// Example: STK Push (Lipa Na M-Pesa Online)
//
// Sends a USSD prompt to a customer's phone to authorize a payment.
// The response only confirms the request was accepted; the actual
// transaction result arrives via the CallBackURL.
//
// Requires: MPESA_CONSUMER_KEY, MPESA_CONSUMER_SECRET, MPESA_PASSKEY
// Verified against the sandbox.
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/yourdudeken/daraja/sdks/go/client"
	"github.com/yourdudeken/daraja/sdks/go/types"
)

func main() {
	mpesa := client.NewClient(types.MpesaConfig{
		ConsumerKey:    os.Getenv("MPESA_CONSUMER_KEY"),
		ConsumerSecret: os.Getenv("MPESA_CONSUMER_SECRET"),
		Environment:    types.Sandbox,
		Passkey:        os.Getenv("MPESA_PASSKEY"),
	})

	ctx := context.Background()

	resp, err := mpesa.STKPush(ctx, types.STKPushRequest{
		BusinessShortCode: 174379,
		TransactionType:   types.CustomerPayBillOnline,
		Amount:            1,
		PartyA:            254708374149, // customer phone sending money
		PartyB:            174379,       // business shortcode receiving money
		PhoneNumber:       254708374149, // phone that receives the STK prompt
		CallBackURL:       "https://example.com/callback",
		AccountReference:  "INV-001",
		TransactionDesc:   "Payment",
	})
	if err != nil {
		log.Fatalf("STK Push failed: %v", err)
	}

	fmt.Printf("CheckoutRequestID: %s\n", resp.CheckoutRequestID)
	fmt.Printf("ResponseCode: %s\n", resp.ResponseCode)
	fmt.Printf("ResponseDescription: %s\n", resp.ResponseDescription)

	// Use CheckoutRequestID with STKQuery to check the payment status.
}
